os/include/tinyara/arch.h

```diff
@@ -387,6 +389,49 @@ FAR void *up_stack_frame(FAR struct tcb_s *tcb, size_t frame_size);
 
 void up_release_stack(FAR struct tcb_s *dtcb, uint8_t ttype);
 
+#ifdef CONFIG_ARMV8M_TRUSTZONE
+/****************************************************************************
+ * Name: up_init_secure_context
+ *
+ * Description:
+ *
+ *   Initializes the secure context management system.
+ *   A task must allocate and load a context before calling any
+ *   secure side function in the thread mode.
+ *
+ ****************************************************************************/
+
+void up_init_secure_context(void);
+
+/****************************************************************************
+ * Name: up_allocate_secure_context
+ *
+ * Description:
+ *    Allocates a context on the secure side.
+ *    It allocates an Opaque context handle if context is
+ *    successfully allocated, NULL otherwise
+ *
+ * Inputs:
+ *   size: Size of the stack to allocate on secure side.
+ *
+ ****************************************************************************/
+
+void up_allocate_secure_context(TZ_ModuleId_t size);
```

Comment:
Could you let me know who calls this?
I found _exit calls up_free_secure_context. That is meaningful when it is allocated.  But I can't find the allocation.

Reply:
I allocates secure context, and is required only if a task wants to call secure world API. In the current code, we call it from rdp demo

Reply:
So you mean that user should call allocation function when user needs. Then, in my opinion, it is the same as malloc and free. How do you think that user should call free function as well?

Reply:
I think allocation is not default, not all threads can call Secure APIs. The free can be handled by the tasks which allocated secure context, however, in case the task is killed this should be taken care by the RTOS. Since the Secure Context handle can be maintained in the tcb, I prefer RTOS to take care of such situations.
I cross-checked other RTOSes, most of which provides task delete() API which applications invokes when it has to end/terminate. The secure context free is handled here. Equivalent to that, I think task_exit() should handle this in TizenRT.

Reply:
That's the same as memory allocation, mm. So, there are two options.

1. your suggestion
pros : no leak with cancel or terminate API
cons : increase tcb size, require checking at every task_exit
2. my suggestion
pros : not require tcb space, condition at task_exit
cons : could have a leak

In my opinion, the thread which uses secure is much small.
For that, we should add the space at tcb structure and should check that at every task_exit even it is not used at almost thread. Usually when we call allocation funtion, we think we should call free as well (It could be my private opinion). To prevent the leak problem, user could use onexit, atexit functionalities.

I really understand what you say. I also have the same thought. But I don't know it is better on system. We need to discuss it more.